hermes-client/src/main/java/pl/allegro/tech/hermes/client/HermesClient.java

```diff
@@ -128,4 +146,39 @@ public void close(long pollInterval, long timeout) throws InterruptedException,
         }
     }
 
+    private void handleMaxRetriesExceeded(ExecutionCompletedEvent<HermesResponse> event) {
+        if (event.getResult().isSuccess()) {
+            return;
+        }
+
+        HermesMessage message = event.getResult().getHermesMessage();
+        if (shouldPublishMetrics()) {
+            String prefix = MetricsUtils.getMetricsPrefix(message.getTopic());
+            metrics.counterIncrement(prefix + ".retries.exhausted");
+        }
```

Comment:
IMO it would be nice to have this metric-specific logic hidden in a `HermesClientMetrics` (or similar) class, which would have the `MetricsProvider` inside. We'd call `metrics.incrementExhausted(message.getTopic())`, `metrics.incrementRetries(message.getTopic())`, etc. There would be no calls for `shouldPublishMetrics()` here.
Probably some classes could be package protected, like `MetricsUtils`.

Reply:
take a look at the latest version, I change it a little bit

Reply:
it looks like HermesClient could now accept an interface - `MessageDeliveryListener`, by default it could accept a `NoopListener`, and the metered one could accept a `MessageDeliveryMeteredListener`. I would prefer it over client inheritance. What do you think? :)

Reply:
I'm not sure how it'll work with the metrics sender though, anyway, I leave it up to you, please decide whether or not it is worth refactoring. :)

Reply:
I agree with You ;-)  I made some changes and I think that's what it was about

Reply:
Wonderful, better than what I suggested. :)